File: src/main/java/uk/yermak/audiobookconverter/MediaInfoProxy.java
package uk.yermak.audiobookconverter;

import java.util.Objects;
import java.util.concurrent.Future;

public class MediaInfoProxy implements MediaInfo {
    private final String filename;
    private final Future<MediaInfo> futureLoad;

    private MediaInfo getMediaInfo() {
        try {
            return this.futureLoad.get();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public int getChannels() {
        return this.getMediaInfo().getChannels();
    }

    public int getFrequency() {
        return this.getMediaInfo().getFrequency();
    }

    public int getBitrate() {
        return this.getMediaInfo().getBitrate();
    }

    @Override
    public String getTitle() {
        return this.getMediaInfo().getTitle();
    }

    @Override
    public String getDetails() {
        return this.getMediaInfo().getDetails();
    }

    public long getDuration() {
        return this.getMediaInfo().getDuration();
    }

    @Override
    public boolean split() {
        return this.getMediaInfo().split();
    }

    @Override
    public void remove() {
        this.getMediaInfo().remove();
    }

    @Override
    public void moveUp() {
        this.getMediaInfo().moveUp();
    }

    @Override
    public void moveDown() {
        this.getMediaInfo().moveDown();
    }

    public String getFileName() {
        return this.filename;
    }

    public AudioBookInfo getBookInfo() {
        return this.getMediaInfo().getBookInfo();
    }

    public ArtWork getArtWork() {
        return this.getMediaInfo().getArtWork();
    }

    public String toString() {
        return this.filename;
    }

    public String getCodec() {
        return this.getMediaInfo().getCodec();
    }

    @Override
    public void setChapter(Chapter chapter) {
        this.getMediaInfo().setChapter(chapter);
    }

    @Override
    public long getOffset() {
        return this.getMediaInfo().getOffset();
    }


    public void setDuration(long duration) {
        this.getMediaInfo().setDuration(duration);
    }

    MediaInfoProxy(final String filename, final Future<MediaInfo> futureLoad) {
        this.filename = filename;
        this.futureLoad = futureLoad;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof MediaInfo)) return false;
        MediaInfo that = (MediaInfo) o;
        return getFileName().equals(that.getFileName());
    }

    @Override
    public int hashCode() {
        return Objects.hash(getFileName(), getDuration()) ;
    }
}

        

File: src/main/scala/uk/yermak/audiobookconverter/MediaInfoProxy.scala
package uk.yermak.audiobookconverter

import java.util.concurrent.{ExecutionException, Future}

/**
  * Created by Yermak on 03-Jan-18.
  */
class MediaInfoProxy(val filename: String, val futureLoad: Future[MediaInfo]) extends MediaInfo {
  private def getMediaInfo = try futureLoad.get
  catch {
    case e@(_: InterruptedException | _: ExecutionException) =>
      e.printStackTrace()
      throw new RuntimeException(e)
  }

  override def setChannels(channels: Int): Unit = getMediaInfo.setChannels(channels)

  override def setFrequency(frequency: Int): Unit = getMediaInfo.setFrequency(frequency)

  override def setBitrate(bitrate: Int): Unit = getMediaInfo.setBitrate(bitrate)

  override def setDuration(duration: Long): Unit = getMediaInfo.setDuration(duration)

  override def getChannels: Int = getMediaInfo.getChannels

  override def getFrequency: Int = getMediaInfo.getFrequency

  override def getBitrate: Int = getMediaInfo.getBitrate

  override def getDuration: Long = getMediaInfo.getDuration

  override def getFileName: String = filename

  override def setBookInfo(bookInfo: AudioBookInfo): Unit = getMediaInfo.setBookInfo(bookInfo)

  override def getBookInfo: AudioBookInfo = getMediaInfo.getBookInfo

  override def getArtWork: ArtWork = getMediaInfo.getArtWork

  override def setArtWork(artWork: ArtWork): Unit = getMediaInfo.setArtWork(artWork)

  override def toString: String = filename

  override def getCodec: String = getMediaInfo.getCodec

  override def setCodec(codec: String): Unit = getMediaInfo.setCodec(codec)
}